crates/shared/src/ethrpc/multicall.rs

```diff
@@ -0,0 +1,426 @@
+//! Multicall encoding and decoding.
+
+use crate::ethrpc::dummy;
+use contracts::support::Multicall;
+use ethcontract::{
+    errors::ExecutionError,
+    tokens::{self, Tokenize as _},
+};
+use hex_literal::hex;
+use lazy_static::lazy_static;
+use std::iter;
+use web3::{
+    self,
+    api::Eth,
+    ethabi::{self, ParamType, Token},
+    types::{AccessList, BlockId, Bytes, CallRequest, H160, U256, U64},
+    Transport,
+};
+
+/// A single call in a multicall batch.
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct Call {
+    /// The address to call.
+    pub to: H160,
+
+    /// Optional gas limit to use for the call.
+    pub gas: Option<U256>,
+
+    /// Optional value to use for the call.
+    pub value: Option<U256>,
+
+    /// Data to use for the call.
+    pub data: Bytes,
+}
+
+impl Call {
+    /// Encode into a tuple used for ABI encoding.
+    pub fn encode(self) -> (H160, U256, U256, ethcontract::Bytes<Vec<u8>>) {
+        (
+            self.to,
+            self.gas.unwrap_or_default(),
+            self.value.unwrap_or_default(),
```

Comment:
Hmm, good point - I should have thought of this a bit more. I think I just blindly mimicked the `web3::CallRequest` API and didn't really think it through.

- For `gas`, it should be:
    ```rust
    enum Gas {
        Auto,
        Exact(NonZeroU256),
    }
    ```
    Since 0 is a "magic" value which means "use all the gas", and everything else is an exact gas limit for the multicall entry. I don't like magic values and I didn't want to introduce a `NonZeroU256`, so I thought `Option<U256>` was a good compromise. OK with changing it if its not clear, or adding clarifications somewhere.
- `value` should totally be `U256`. `None` means no value, so there is no reason that it should be an option. Just copy-past from the `web3::CallRequest` type.

Reply:
I changed `value: U256`, since `Option` didn't really make sense.

I kept `gas: Option<256>` for now, let me know if my explanation ☝️ made sense and whether or not you have any suggestions.